src/Channels/Memory.cs

```diff
@@ -0,0 +1,90 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+namespace Channels
+{
+    public struct Memory<T>
+    {
+        private readonly T[] _array;
+        private readonly int _offset;
+        private readonly unsafe void* _memory;
+        private readonly int _memoryLength;
+
+        public unsafe Memory(void* pointer, int offset, int length)
+        {
+            unsafe
+            {
+                _memory = pointer;
+            }
+
+            _array = null;
+            _offset = offset;
+            _memoryLength = length;
+        }
+
+        public unsafe Memory(T[] array, int offset, int length, void* pointer = null)
+        {
+            unsafe
+            {
+                _memory = pointer;
+            }
+
+            _array = array;
+            _offset = offset;
+            _memoryLength = length;
+        }
+
+        public Span<T> Span => this;
+
+        public bool IsEmpty => Length == 0;
+
+        public static implicit operator Span<T>(Memory<T> memory)
+        {
+            if (memory.Length == 0)
+            {
+                return Span<T>.Empty;
+            }
+
+            if (memory._array != null)
+            {
+                return memory._array.Slice(memory._offset, memory.Length);
+            }
+            else
+            {
+                unsafe
+                {
+                    return new Span<T>(memory._memory, memory._memoryLength);
+                }
+            }
+        }
+
+        public unsafe void* UnsafePointer => (byte*)_memory + (Unsafe.SizeOf<T>() * _offset);
```

Comment:
@davidfowl if the question there is "should it, in the "`_memory` is nil" case, itself fetch the address from the array", then IMO yeah, it feels like they should be accessing the array and using `fixed`, because *you can't know* that they have it pinned. However, the more immediate concern is that currently, in the "`_memory` is nil" case, this will return gibberish but non-zero gibberish. If we're lucky it'll still cause a memory access violation; if we're unlucky, they have a pointer to a random place in the process's memory.